File: backend/app/model/domain.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

Protocol = Literal["TCP", "UDP", "ICMP", "OTHER"]


class TrafficEvent(BaseModel):
    ts: float
    srcIp: str
    dstIp: str
    protocol: Protocol
    port: int | None = None
    bytes: int = Field(ge=0)


class ScoreBreakdown(BaseModel):
    geoSpread: float
    destinationReputation: float
    contactFrequency: float
    knownTracker: float
    finalScore: int = Field(ge=0, le=100)


class Device(BaseModel):
    id: str
    ip: str
    mac: str | None = None
    hostname: str | None = None
    vendor: str | None = None
    firstSeen: float
    lastSeen: float
    bytesIn: int = 0
    bytesOut: int = 0
    trustScore: int = Field(ge=0, le=100, default=100)
    scoreBreakdown: ScoreBreakdown = Field(
        default_factory=lambda: ScoreBreakdown(
            geoSpread=0.0,
            destinationReputation=0.0,
            contactFrequency=0.0,
            knownTracker=0.0,
            finalScore=100,
        )
    )


class Connection(BaseModel):
    id: str
    srcDeviceId: str
    dstIp: str
    dstHost: str | None = None
    protocol: Protocol
    port: int | None = None
    bytes: int = 0
    packetCount: int = 0
    lastSeen: float

File: frontend/src/types/domain.ts
export type Protocol = "TCP" | "UDP" | "ICMP" | "OTHER";

export interface TrafficEvent {
  ts: number;
  srcIp: string;
  dstIp: string;
  protocol: Protocol;
  port: number | null;
  bytes: number;
}

export interface ScoreBreakdown {
  geoSpread: number;
  destinationReputation: number;
  contactFrequency: number;
  knownTracker: number;
  finalScore: number;
}

export interface Device {
  id: string;
  ip: string;
  mac: string | null;
  hostname: string | null;
  vendor: string | null;
  firstSeen: number;
  lastSeen: number;
  bytesIn: number;
  bytesOut: number;
  trustScore: number;
  scoreBreakdown: ScoreBreakdown;
}

export interface Connection {
  id: string;
  srcDeviceId: string;
  dstIp: string;
  dstHost: string | null;
  protocol: Protocol;
  port: number | null;
  bytes: number;
  packetCount: number;
  lastSeen: number;
}
